### ISSUE TITLE
All packages are marked as unused if project uses external autoload

### ISSUE THREAD

**Issue by lulco** [composer-unused/composer-unused#119]
Hi @icanhazstring :) it's me again :)

## Describe the bug
I just test composer-unused on some real app (not library) and I get interesting results.
With `composer-unused -x` all packages are marked as unused. Without `-x` option there are some used and ignored packages.

## Additional information
I realized that this project has no autoload in composer.json because it uses some external loader. So when I add
```
"autoload" : {
    "classmap": ["app"]
}
```
results were much better. So it seems in version with `-x` option the autoload is required to be in composer.json but in version without `-x` option it works without it (but of course packages with autoload classmap are ignored). So we need some combination of these two approaches O:-)

###
composer-unused version: 0.8.x-dev
tested on php: 7.3.8 and 7.4.9

**Comment by icanhazstring** [comment]
Hi @lulco

Could you paste the full output of `composer-unused` with and without the `-x` option?

Also could you give me the `composer.json` of the app. You can ofc remove some proprietary dependencies if you need to.

**Comment by lulco** [comment]
Sure, I had to create some minimal repository. I again used my old small app.

`composer.json`
```
{
    "require": {
        "php": ">= 7.3",
        "ext-json": "*",
        "lulco/redis-proxy": "^0.4.2",
        "tomaj/nette-api": "^2.0",
        "nette/database": "^3.0",
        "lulco/phoenix": "^1.3",
        "predis/predis": "^1.1"
    }
}
```

App contains 2 classes:

`app/Api/MyApiHandler.php`:
```
<?php

namespace ComposerUnusedTest\Api;

use Nette\Database\Context;
use Nette\Http\IResponse;
use Tomaj\NetteApi\Handlers\BaseHandler;
use Tomaj\NetteApi\Response\JsonApiResponse;
use Tomaj\NetteApi\Response\ResponseInterface;

class MyApiHandler extends BaseHandler
{
    /** @var Context */
    private $context;

    public function __construct(Context $context)
    {
        parent::__construct();
        $this->context = $context;
    }

    public function handle(array $params): ResponseInterface
    {
        return new JsonApiResponse(IResponse::S200_OK, []);
    }
}
```
and `app/Api/MyApiHandler.php`:
```
<?php

namespace ComposerUnusedTest\Redis;

use RedisProxy\RedisProxy;

class MyRedis
{
    /** @var RedisProxy */
    private $redisProxy;

    public function __construct(RedisProxy $redisProxy)
    {
        $this->redisProxy = $redisProxy;
    }
}
```

Output for `composer-unused -x`:
```
Results
-------

Found 0 used, 7 unused and 0 ignored packages

 Used packages

 Unused packages
 ✗ php
 ✗ ext-json
 ✗ lulco/redis-proxy
 ✗ tomaj/nette-api
 ✗ nette/database
 ✗ lulco/phoenix
 ✗ predis/predis

 Ignored packages
```

Output for `composer-unused`:
```
Loading packages
----------------

 7/7 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%

 ! [NOTE] Found 4 package(s) to be checked.                                                                             

Scanning files from basedir /var/www/composer-unused-test
---------------------------------------------------------

 2/2 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%

Results
-------

Found 3 used, 1 unused and 3 ignored packages

 Used packages
 ✓ php (required by: tomaj/nette-api)
 ✓ ext-json (required by: tomaj/nette-api)
 ✓ tomaj/nette-api

 Unused packages
 ✗ predis/predis

 Ignored packages
 ○ lulco/redis-proxy (Package provides no namespace)
 ○ nette/database (Package provides no namespace)
 ○ lulco/phoenix (Package provides no namespace)
```



With changed composer.json (added autoload classmap app):
```
{
    "require": {
        "php": ">= 7.3",
        "ext-json": "*",
        "lulco/redis-proxy": "^0.4.2",
        "tomaj/nette-api": "^2.0",
        "nette/database": "^3.0",
        "lulco/phoenix": "^1.3",
        "predis/predis": "^1.1"
    },
    "autoload": {
        "classmap": ["app"]
    }
}
```

Output for `composer-unused -x` is as expected:

```
Results
-------

Found 6 used, 1 unused and 0 ignored packages

 Used packages
 ✓ php
 ✓ ext-json
 ✓ lulco/redis-proxy
 ✓ tomaj/nette-api
 ✓ nette/database
 ✓ predis/predis

 Unused packages
 ✗ lulco/phoenix

 Ignored packages
```

**Comment by icanhazstring** [comment]
Thanks, will take a look into it 👍

**Comment by icanhazstring** [comment]
Ok, found the problem.

With `0.7.x` I use the basepath of the `composer.json` to recursively scan all `*.php` files. So the configuration about the root autoload is irrelevant.

With `0.8.x` I used the `composer.autoload` configuration. So if that is not present, nothing will be loaded, resulting in no used symbols. That is ofc, "stupid" ;)

In case of `composer.autoload` not present I should have a fallback of recursively scanning the root folder of the project. Which might ofc result in a longer execution time. Maybe I'll add a notice if that is case.

**Comment by icanhazstring** [comment]
Hi @lulco this is fix in `dev-main` soon to be released with `0.8`

**Comment by lulco** [comment]
Hi @icanhazstring it's been a while :) I stopped using the package, but maybe I will try it again soon :) Thanks
